xdsl/transforms/riscv_scf_loop_range_folding.py

```diff
@@ -0,0 +1,95 @@
+from xdsl.dialects import builtin, riscv, riscv_scf
+from xdsl.ir.core import MLContext
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+
+
+class HoistIndexTimesConstant(RewritePattern):
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: riscv_scf.ForOp, rewriter: PatternRewriter) -> None:
+        index = op.body.block.args[0]
+
+        uses: list[riscv.MulOp] = []
+        constant: int | None = None
+        for use in index.uses:
+            if not isinstance(mul_op := use.operation, riscv.MulOp):
+                # One of the uses is not a multiplication, bail
+                return
+
+            uses.append(mul_op)
+
+            if mul_op.rs1 is index:
+                if not isinstance(li_op := mul_op.rs2.owner, riscv.LiOp):
+                    # One of the uses is not a multiplication by constant, bail
+                    return
+            else:
+                if not isinstance(li_op := mul_op.rs1.owner, riscv.LiOp):
+                    # One of the uses is not a multiplication by constant, bail
+                    return
+
+            if not isinstance(imm := li_op.immediate, builtin.IntegerAttr):
+                # One of the constants is not an integer, bail
+                return
+
+            if constant is None:
+                constant = imm.value.data
+            else:
+                if constant != imm.value.data:
+                    # Not all constants are equal, bail
+                    return
+
+        if constant is None:
+            # No uses, bail
+            return
+
+        # All the uses are multiplications by a constant, we can fold
+        rewriter.insert_op_before_matched_op(
+            [
+                factor := riscv.LiOp(constant),
+                new_lb := riscv.MulOp(
+                    op.lb, factor, rd=riscv.IntRegisterType.unallocated()
+                ),
+                old_distance := riscv.SubOp(
+                    op.ub, op.lb, rd=riscv.IntRegisterType.unallocated()
+                ),
+                new_distance := riscv.MulOp(
+                    old_distance, factor, rd=riscv.IntRegisterType.unallocated()
+                ),
+                new_ub := riscv.AddOp(
+                    op.lb, new_distance, rd=riscv.IntRegisterType.unallocated()
+                ),
+                new_ub := riscv.AddOp(
+                    new_lb, new_distance, rd=riscv.IntRegisterType.unallocated()
+                ),
+                new_step := riscv.MulOp(
+                    op.step, factor, rd=riscv.IntRegisterType.unallocated()
+                ),
```

Comment:
I'm afraid I don't quite understand why we need this "dance" here with distances, etc. Shouldn't we just be able to go: `new_x = x * const` for lb, ub and step?

Also, it may or may not be a good idea to do overflow checking?

Reply:
Can you come up with the counterexample? ;)

Reply:
I no longer remember why I added it, so I'll remove it

Reply:
The overflow check sounds like a good idea, not sure if it'll ever be a practical concern